src/Microsoft.AspNetCore.Routing/Matching/RefEmitTrieJumpTable.cs

```diff
@@ -0,0 +1,167 @@
+
+                il.Emit(OpCodes.Ldarg_2);
+                il.Emit(OpCodes.Ldc_I4, group.Key);
+                il.Emit(OpCodes.Beq, nextGroup);
+
+                // Process the group
+                EmitTableForStringsOfEqualLength(il, group.ToArray(), 0, group.Key, offset, c, returnDefault);
+            }
+
+            il.MarkLabel(nextGroup);
+            il.MarkLabel(returnDefault);
+            il.Emit(OpCodes.Ldc_I4, _defaultDestination);
+            il.Emit(OpCodes.Ret);
+
+#if NET462
+            var t = type.CreateType();
+            assembly.Save(assemblyName.Name + ".dll");
+            return (Func<string, int, int, int>)t.GetMethod("GeneratedGetDestination").CreateDelegate(typeof(Func<string, int, int, int>));
+#else
+            return (Func<string, int, int, int>)method.CreateDelegate(typeof(Func<string, int, int, int>));
+#endif
+        }
+
+        private static void EmitTableForStringsOfEqualLength(
+            ILGenerator il,
+            (string text, int destination)[] entries,
+            int index,
+            int length,
+            LocalBuilder offset,
+            LocalBuilder c,
+            Label returnDefault)
+        {
+            // We've reached the end of the string.
+            if (index == length)
+            {
+                Debug.Assert(entries.Length == 1, "We should have a single entry");
+                il.Emit(OpCodes.Ldc_I4, entries[0].destination);
+                il.Emit(OpCodes.Ret);
+                return;
+            }
+
+            // offset++
+            il.Emit(OpCodes.Ldloc, offset);
+            il.Emit(OpCodes.Ldc_I4, 1);
+            il.Emit(OpCodes.Add);
+            il.Emit(OpCodes.Stloc, offset);
+
+            // c = path[offset]
+            il.Emit(OpCodes.Ldarg_1);
+            il.Emit(OpCodes.Ldloc, offset);
+            il.Emit(OpCodes.Callvirt, typeof(string).GetProperty("Chars", new Type[] { typeof(int), }).GetMethod);
+            il.Emit(OpCodes.Stloc, c);
+
+            var nextGroup = il.DefineLabel();
+            var groups = entries.GroupBy(e => char.ToUpperInvariant(e.text[index])).ToArray();
+            for (var i = 0; i < groups.Length; i++)
+            {
+                var group = groups[i];
+
+                // Similar to 'if (c == 'A' || c == 'a') { ... }
+                il.MarkLabel(nextGroup);
+                nextGroup = il.DefineLabel();
+
+                il.Emit(OpCodes.Ldloc, c);
+                il.Emit(OpCodes.Ldc_I4, group.Key);
+                il.Emit(OpCodes.Beq, nextGroup);
+
+                if (char.ToLowerInvariant(group.Key) != group.Key)
+                {
+                    il.Emit(OpCodes.Ldloc, c);
+                    il.Emit(OpCodes.Ldc_I4, char.ToLowerInvariant(group.Key));
+                    il.Emit(OpCodes.Beq, nextGroup);
+                }
+
+                // Process the group
+                EmitTableForStringsOfEqualLength(il, group.ToArray(), index + 1, length, offset, c, returnDefault);
```

Comment:
If I'm reading this correctly, this is grouping every character through to the end of comparison. At some point when there is just one match possible remaining can the generated IL be simplified?

Reply:
yes, we could absolutely do something like that. It would end up looking at lot like https://github.com/aspnet/Routing/pull/729

Reply:
I'm not sure it's worth it, both in terms of performance and effort. One idea I'd like to try is to see if this implementation is faster for techempower plaintext. Then we might end up with a single 'fast path'.